subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java

```diff
@@ -0,0 +1,304 @@
+    private final ImmutableList.Builder<CachingDisabledReason> noCachingReasonsBuilder = ImmutableList.builder();
+
+    @Override
+    public void appendImplementation(ImplementationSnapshot implementation) {
+        this.implementation = implementation;
+        if (implementation.isUnknown()) {
+            noCachingReasonsBuilder.add(new CachingDisabledReason(
+                CachingDisabledReasonCatwgory.NON_CACHEABLE_IMPLEMENTATION,
+                "Implementation type " + implementation.getUnknownReason()
+            ));
+        }
+    }
+
+    @Override
+    public void appendAdditionalImplementation(ImplementationSnapshot additionalImplementation) {
+        this.additionalImplementationsBuilder.add(additionalImplementation);
+        if (additionalImplementation.isUnknown()) {
+            noCachingReasonsBuilder.add(new CachingDisabledReason(
+                CachingDisabledReasonCatwgory.NON_CACHEABLE_ADDITIONAL_IMPLEMENTATION,
+                "Additional implementation type " + additionalImplementation.getUnknownReason()
+            ));
+        }
+    }
+
+    @Override
+    public void appendInputValueFingerprint(String propertyName, ValueSnapshot fingerprint) {
+        Hasher hasher = Hashing.newHasher();
+        fingerprint.appendToHasher(hasher);
+        if (hasher.isValid()) {
+            HashCode hash = hasher.hash();
+            recordInputValueFingerprint(propertyName, hash);
+        } else {
+            markInputValuePropertyNotCacheable(propertyName, hasher.getInvalidReason());
+        }
+    }
+
+    protected void recordInputValueFingerprint(String propertyName, HashCode fingerprint) {
+        inputValueFingerprintsBuilder.put(propertyName, fingerprint);
+    }
+
+    protected void markInputValuePropertyNotCacheable(String propertyName, String nonCacheableReason) {
+        nonCacheableInputPropertiesBuilder.put(propertyName, nonCacheableReason);
+    }
+
+    @Override
+    public void appendInputFilesPropertyFingerprints(String propertyName, CurrentFileCollectionFingerprint fingerprints) {
+        inputFileFingerprintsBuilder.put(propertyName, fingerprints);
+    }
+
+    @Override
+    public void appendOutputPropertyName(String propertyName) {
+        outputPropertiesBuilder.add(propertyName);
+    }
+
+    @Override
+    public void markNotCacheable(CachingDisabledReason reason) {
+        noCachingReasonsBuilder.add(reason);
+    }
+
+    @Override
+    public CachingState build() {
+        ImmutableList<ImplementationSnapshot> additionalImplementations = additionalImplementationsBuilder.build();
+        ImmutableSortedMap<String, HashCode> inputValueFingerprints = inputValueFingerprintsBuilder.build();
+        ImmutableSortedMap<String, CurrentFileCollectionFingerprint> inputFileFingerprints = inputFileFingerprintsBuilder.build();
+        ImmutableSortedSet<String> outputProperties = outputPropertiesBuilder.build();
+
+        Hasher hasher = Hashing.newHasher();
```

Comment:
If we have some invalid input values, we are calculating a partial cache key here. Does this make sense? Shouldn't we just not calculate a cache key if there are some invalid inputs?

Reply:
Right, that's one thing I wanted to do, i.e. move the hasher invalidation out from `Hasher`.

Reply:
Actually, when there are invalid input values, then we don't calculate the hash. We could stop feeding the hasher, too, but that doesn't sound like a valuable optimization.

I tried to make the code a bit more clear.